### PR TITLE
wip: Mgmt 3362 kubeapi in subsystem

### PR DESCRIPTION
**Pull request by RazRegev**

### CODE REVIEW

**Comment by RazRegev** [comment]
/cc @filanov 
This is still in progress but I will appreciate it if you could take a look and tell me what you think of the general idea.

File: subsystem/apiclient/kubeapi.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+func (kc *KubeAPIClient) RegisterCluster(
+	ctx context.Context,
+	params *installer.RegisterClusterParams,
+) (*models.Cluster, error) {
+
+	cParams := params.NewClusterParams
+
+	clusterName := swag.StringValue(cParams.Name)
+
+	pullSecretRef, pullSecretErr := kc.deployPullSecret(
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:62]: Maybe instead of creating a new pull secret for every cluster we can create one global pull secret on the setup of the batch? we have the init function in  subsystem_suite_test.go

File: subsystem/apiclient/kubeapi.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+			UserManagedNetworking:    swag.BoolValue(cParams.UserManagedNetworking),
+			AdditionalNtpSource:      swag.StringValue(cParams.AdditionalNtpSource),
+			PullSecretRef:            pullSecretRef,
+		},
+	}
+	if deployErr := kc.client.Create(ctx, &c); deployErr != nil {
+		return nil, errors.Wrapf(deployErr, "failed to deploy cluster")
+	}
+
+	key := types.NamespacedName{
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:102]: All the get part should probably be a helper function.

File: subsystem/apiclient/kubeapi.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+	secret = &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: kc.namespace,
+		},
+		Data: secretData,
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:150]: I think that you can use `StringData` and then you will not need to encrypt.
it doesn't work in unit-tests, but i did worked for me in minikube.

File: subsystem/apiclient/kubeapi.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	c := adiiov1alpha1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Cluster",
+			APIVersion: APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metaName,
+			Namespace: kc.namespace,
+		},
+		Spec: adiiov1alpha1.ClusterSpec{
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:258]: I think that some fields are missing here, for example `MachineNetworkCIDR` etc.

File: subsystem/apiclient/kubeapi.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+		}
+		break
+	}
+	if getClusterErr != nil {
+		return nil, errors.Wrapf(getClusterErr, "failed to get cluster")
+	}
+	return cluster, nil
+}
+
+func (kc *KubeAPIClient) getClusterByKey(key types.NamespacedName) (*models.Cluster, error) {
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:193]: I'm not sure if we should take the cluster directly from the DB or get it with KubeAPI and translate to `model.Cluster` this way we will hide translation bugs. but it can probably be made in a different PR.

File: subsystem/apiclient/kubeapi.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+			UserManagedNetworking:    swag.BoolValue(cParams.UserManagedNetworking),
+			AdditionalNtpSource:      swag.StringValue(cParams.AdditionalNtpSource),
+			PullSecretRef:            pullSecretRef,
+		},
+	}
+	if deployErr := kc.client.Update(ctx, &c); deployErr != nil {
+		return nil, errors.Wrapf(deployErr, "failed to deploy cluster")
+	}
+
+	key := types.NamespacedName{
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:279]: Probably need to take kube resource and check that t here are no errors in the update. `cluster.Status.Error`
Currently we have issues with some of the update parameters [MGMT-3598 ](https://issues.redhat.com/browse/MGMT-3598) but still the right validation will be to see that we don't have any issues.

File: subsystem/apiclient/kubeapi.go [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+		Name:      metaName,
+		Namespace: kc.namespace,
+	}
+	cluster, getClusterErr := kc.getClusterByKey(key)
```
Comment (filanov) [inline subsystem/apiclient/kubeapi.go:283]: I think that there maybe a race here, we don't really know when the resource was updated because the API is async. 
We can take a look at `updated_at` but it's not perfect because cluster can be updated by the monitors as well.
@mhrivnak how can we validate that reconcile was called for specific resource?

**Comment by filanov** [comment]
@RazRegev I'm suggesting to add a simple test that validate existing apis. 
I saw that there is not cleanup for the CRs, maybe it should be added to clear-db or something like that. at the end of the sute you can delete all the resources including pull-secret.

File: subsystem/apiclient/kubeapi.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+const (
+	APIVersion = "adi.io.my.domain/v1alpha1"
```
[the remark this answers is not in the document]
Comment (dgoodwin) [reply subsystem/apiclient/kubeapi.go:26]: Will become hive.openshift.io/v1 and use ClusterDeployment as soon as we can get https://github.com/openshift/hive/pull/1247 settled and merged.

[the remark this answers is not in the document]
Comment (filanov) [reply subsystem/apiclient/kubeapi.go:26]: Need to remove the domain.

File: subsystem/apiclient/kubeapi.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+	if err != nil {
+		return nil, err
+	}
+	kc := &KubeAPIClient{
+		client:    client,
+		namespace: namespace,
+		db:        db,
+	}
+	return kc, nil
+}
```
Comment (dgoodwin) [inline subsystem/apiclient/kubeapi.go:51]: What is the subsystem layer in your code base?

Curious because your controllers have a client already, so I wasn't sure what the purpose of creating another was with this utility.

File: subsystem/apiclient/kubeapi.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+	pullSecretRef, pullSecretErr := kc.deployPullSecret(
+		ctx,
+		clusterName,
+		swag.StringValue(cParams.PullSecret))
+	if pullSecretErr != nil {
+		return nil, errors.Wrapf(pullSecretErr, "failed to deploy pull secret")
+	}
+
+	metaName := uuid.New().String()
```
Comment (dgoodwin) [inline subsystem/apiclient/kubeapi.go:70]: Should this just be "clusterName"? We are most likely targeting one cluster per namespace, not much risk of collision. But as with previous comment I'm slightly confused what this layer is as I'd expect this Kube CR to be provided by a user, and most likely with a more friendly meta name.

File: subsystem/apiclient/restapi.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+) error {
+
+	_, err := rc.ai.Installer.DeregisterCluster(ctx, params)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (rc *RestAPIClient) UpdateCluster(
```
Comment (dgoodwin) [inline subsystem/apiclient/restapi.go:44]: Noticed this was missing on your APIclient interface.

**Review by dgoodwin — COMMENTED** [review]
I might need some more context but just did a quick scan and added a couple thoughts. Sorry if I can be of more help! Most of all curious what this subsystem layer is.

File: subsystem/apiclient/kubeapi.go (hunk 9 above)
Comment (mhrivnak) [reply subsystem/apiclient/kubeapi.go:51]: It's a test suite whose scope is bigger than unit tests but smaller than e2e tests, generally targeting one "subsystem" at a time.

**Comment by mhrivnak** [comment]
Am I understanding correctly that you're trying to run the same exact tests, but use a different client that will either talk to the rest api or use CRDs depending on how it's configured? I don't think that's going to be a successful endeavor. The imperative vs. declarative workflow is very different from a client perspective. I don't see how we could have a 1-to-1 mapping of actions. The question that's already been raised about the race with the update call is a good example, and just the beginning.

File: subsystem/apiclient/kubeapi.go (hunk 7 above)
Comment (mhrivnak) [reply subsystem/apiclient/kubeapi.go:283]: I made a general comment about this: https://github.com/openshift/assisted-service/pull/923#issuecomment-761181513

**Comment by filanov** [comment]
> Am I understanding correctly that you're trying to run the same exact tests, but use a different client that will either talk to the rest api or use CRDs depending on how it's configured? I don't think that's going to be a successful endeavor. The imperative vs. declarative workflow is very different from a client perspective. I don't see how we could have a 1-to-1 mapping of actions. The question that's already been raised about the race with the update call is a good example, and just the beginning.

I'll setup a meeting i think that after we clear things up we can talk about possible solutions.

File: subsystem/apiclient/kubeapi.go [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+			SSHPublicKey:             swag.StringValue(&cParams.SSHPublicKey),
+			VIPDhcpAllocation:        swag.BoolValue(cParams.VipDhcpAllocation),
+			HTTPProxy:                swag.StringValue(cParams.HTTPProxy),
+			HTTPSProxy:               swag.StringValue(cParams.HTTPSProxy),
+			NoProxy:                  swag.StringValue(cParams.NoProxy),
+			UserManagedNetworking:    swag.BoolValue(cParams.UserManagedNetworking),
+			AdditionalNtpSource:      swag.StringValue(cParams.AdditionalNtpSource),
+			PullSecretRef:            pullSecretRef,
+		},
+	}
```
[the remark this answers is not in the document]
Comment (RazRegev) [reply subsystem/apiclient/kubeapi.go:92]: this client will be used by our tests, this code won't be part of the service itself

**Comment by RazRegev** [comment]
will open again
